### PR TITLE
Add "add_points" method for point/symbol data

### LINKED ISSUE #36: Add points to an image with list of lat/long pairs
Although the `add_cities` function can overlay points on an image, but a Shapefile describing the points is required for this function.
Referring to [Custom shapes and lines](https://pycoast.readthedocs.io/en/latest/polygons_and_lines.html), is there any similar function to `add_polygon` which can overlay points on a PNG image simply by pairs of lat/long coordinates?

### PR DESCRIPTION
**Pull request by yufeizhu600**
- [x] Closes #36 
 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/master **/*py | flake8 --diff`` 
 - [x] Fully documented 
 - [x] AUTHORS.md updated.   
   
This PR implement a new feature to add a symbol and/or text description at one point or a list of points of interest to an image. 

A list of poi can be passed to the add_poi() function with the format [((lon1, lat1), description1), ((lon2, lat2), description2)]; the symbol can be chosen from ['circle', 'square', 'asterisk']; an optional description string can be added next to the symbol on the map.
A quick example as below:
```        
        poi_list = [((2.3522, 48.8566), 'Paris'),
                    ((0.1278, 51.5074), 'London')]
        cw.add_poi(img, area_def, poi_list=poi_list, font_file=font_file,
                   symbol='circle', ptsize=16,
                   outline='black', width=3,
                   fill='red', fill_opacity=128,
                   textbox_outline='blue', textbox_linewidth=0.5,
                   textbox_fill='yellow', textbox_opacity=200)

```
Which will add two red dots at the two locations with the name of the cities. Please find the result images in the unit test directory.

It also add support for add_overlay_from_config() function to load poi parameters from a config file similar as add_coastlines(), add_borders() functions. The example can also be found in the unit test directory.

### CODE REVIEW

File: pycoast/cw_agg.py [hunk 1]
```diff
@@ -61,23 +61,35 @@ def _draw_polygon(self, draw, coordinates, **kwargs):
...
-        pen = aggdraw.Pen(kwargs['outline'])
+        pen = aggdraw.Pen(kwargs['outline'],
+                          kwargs['width'],
+                          kwargs['outline_opacity'])
 
-        fill_opacity = kwargs.get('fill_opacity', 255)
+        if kwargs['fill'] is None:
+            fill_opacity = 0
+        else:
+            fill_opacity = kwargs['fill_opacity']
```
Comment (djhoese) [inline pycoast/cw_agg.py:71]: Why the change in default? And why get rid of `.get`?

File: pycoast/cw_agg.py [hunk 2]
```diff
@@ -61,23 +61,35 @@ def _draw_polygon(self, draw, coordinates, **kwargs):
 
     def _draw_rectangle(self, draw, coordinates, **kwargs):
         """Draw rectangle."""
-        pen = aggdraw.Pen(kwargs['outline'])
+        pen = aggdraw.Pen(kwargs['outline'],
+                          kwargs['width'],
+                          kwargs['outline_opacity'])
```
Comment (djhoese) [inline pycoast/cw_agg.py:66]: How did this ever work before?

File: pycoast/cw_agg.py [hunk 3]
```diff
@@ -101,6 +116,40 @@ def _draw_line(self, draw, coordinates, **kwargs):
                           kwargs['outline_opacity'])
         draw.line(coordinates, pen)
 
+    def _draw_asterisk(self, draw, pt_size, coordinate, **kwargs):
+        """Draw a asterisk sign '*' at the given coordinate """
```
Comment (djhoese) [inline pycoast/cw_agg.py:120]: Missing period at the end?

File: pycoast/cw_agg.py (hunk 1 above)
Comment (yufeizhu600) [reply pycoast/cw_agg.py:71]: I think it is something I played with but forgot to restore it back. Will modify it in next push.

File: pycoast/cw_agg.py (hunk 2 above)
Comment (yufeizhu600) [reply pycoast/cw_agg.py:66]: I think aggdraw have default values for these args

File: pycoast/cw_agg.py (hunk 3 above)
Comment (yufeizhu600) [reply pycoast/cw_agg.py:120]: Thanks, I will add it.

**Comment by yufeizhu600** [comment]
I see your **points**, the plural word makes a cleaner fun name to the users. I will adapt it.  Thanks.

File: docs/source/points.rst [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    >>> from pycoast import ContourWriterAGG
+    >>> img = Image.new('RGB', (1024, 1024), (255, 255, 255))
+    >>> proj4_string = '+proj=laea +lat_0=90 +lon_0=0 +a=6371228.0 +units=m'
+    >>> area_extent = (-5326849.0625, -5326849.0625, 5326849.0625, 5326849.0625)
+    >>> area_def = AreaDefinition('nh', 'nh', 'nh', proj4_string, 1024, 1024, area_extent)
+    >>> cw = ContourWriterAGG('/home/esn/data/gshhs')
+    >>> cw.add_coastlines(img, area_def, outline='black', resolution='l', level=4)
+    >>> cw.add_borders(img, area_def, outline='black', width=3, level=1, resolution='c')
+
+Now we can add a cycle, which is the deault symbol, with default point size 6 at the
```
Comment (djhoese) [inline docs/source/points.rst:20]:
```suggestion
Now we can add a circle, which is the default symbol, with default point size 6 at the
```

File: docs/source/points.rst [hunk 5]
```diff
@@ -0,0 +1,68 @@
+Custom points with or without textbox
+--------------------------------
+
+Pycoast can add a symbol to points of interest on an image. The following examples show how
+we might use the :attr:`add_points` method to anotate the points on an image.
+
+First of all, we setup an PIL image with an area definition, then we add coastlines and
```
Comment (djhoese) [inline docs/source/points.rst:7]:
```suggestion
First of all, we setup a PIL image with an area definition, then we add coastlines and
```

File: docs/source/points.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
+Custom points with or without textbox
+--------------------------------
+
+Pycoast can add a symbol to points of interest on an image. The following examples show how
+we might use the :attr:`add_points` method to anotate the points on an image.
```
Comment (djhoese) [inline docs/source/points.rst:5]:
```suggestion
we might use the :meth:`~pycoast.cw_agg.ContourWriterAGG.add_points` method to annotate the points on an image.
```

File: docs/source/points.rst [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+The example below will add 'Rome' at the given location without a symbol.
+
+    >>> points_list = [((12.4964, 41.9028), 'Rome')]
+    >>> cw.add_points(pil_img, area, points_list=points_list,
+    ...               font_file=font_file, font_size=16,
+    ...               symbol='circle', ptsize=0,
+    ...               textbox_outline='black', text_linewidth=1,
+    ...               textbox_fill='yellow', textbox_opacity=200)
+
+Simularly, assign the descrition as an empty string, will only draw the symbol on the image.
```
Comment (djhoese) [inline docs/source/points.rst:37]:
```suggestion
Similarly, assign the description as an empty string will only draw the symbol on the image.
```

File: docs/source/points.rst [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+The example below will draw a square symbol at the localtion of Paris.
```
Comment (djhoese) [inline docs/source/points.rst:38]:
```suggestion
The example below will draw a square symbol at the location of Paris.
```

File: docs/source/points.rst [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    >>> points_list = [((2.3522, 48.8566), '')]
+    >>> cw.add_points(pil_img, area, points_list=points_list,i
+    ...               font_file=font_file,
+    ...               symbol='square', ptsize=10,
+    ...               outline='red', width=1,
+    ...               fill='blue', fill_opacity=128)
+
+Finally, we can fully costomized the annotation as the example below, which will add
```
Comment (djhoese) [inline docs/source/points.rst:47]:
```suggestion
Finally, we can fully customize the annotation as the example below, which will add
```

File: docs/source/points.rst [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+a circle in black with linewide set to 2 and filled in red color with opacity equals 255;
```
Comment (djhoese) [inline docs/source/points.rst:48]:
```suggestion
a circle in black with linewidth set to 2 and filled in red color with opacity equals 255;
```

File: docs/source/points.rst [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    ...               symbol='circle', ptsize=14,
+    ...               outline='black', width=2,
+    ...               fill='red', fill_opacity=255,
+    ...               textbox_outline='blue', textbox_linewidth=1.5,
+    ...               textbox_fill='green', textbox_opacity=128)
+    >>> img.show()
+
+.. image:: images/nh_points_agg.png
+
+The :attr:`add_points` method accepts a list of longitude, latitude pairs, with an optional
```
Comment (djhoese) [inline docs/source/points.rst:64]:
```suggestion
The :meth:`~pycoast.cw_agg.ContourWriterAGG.add_points` method accepts a list of longitude, latitude pairs, with an optional
```

Comment (djhoese) [reply docs/source/points.rst:64]: This sentence seems left over and kind of out of order with the rest of the documentation. Like you are saying what the method takes as arguments after showing a ton of examples of how to use it. Maybe this should go earlier on this page?

File: pycoast/cw_base.py [hunk 12]
```diff
@@ -924,24 +947,131 @@ def add_cities(self, image, area_def, citylist, font_file, font_size,
...
+                    self._draw_text_box(draw, text_position, city_name, font, outline,
+                                        box_outline, box_opacity)
                     logger.info("%s added", str(city_name))
 
         self._finalize(draw)
 
+    def add_points(self, image, area_def, points_list, font_file, font_size=12,
+                   symbol='circle', ptsize=6, outline='black', fill='white', **kwargs):
+        """Add a symbol and/or text at the point(s) of interest to a PIL image object.
+        :Parameters:
```
Comment (djhoese) [inline pycoast/cw_base.py:970]: If you generate the sphinx docs, does this complain? I'm wondering if there needs to be a blank line above Parameters.

File: pycoast/cw_base.py [hunk 13]
```diff
@@ -924,24 +947,131 @@ def add_cities(self, image, area_def, citylist, font_file, font_size,
...
+            textbox_outline : str or (R, G, B), optional
+                Line color of the textbox borders.
+            textbox_linewidth : float
+                Line width of the the borders of the textbox
+            textbox_fill : str or (R, G, B), optional
+                Filling color of the background of the textbox.
+            textbox_opacity : int, optional {0; 255}
+                Opacity of the background filling of the textbox.
+        """
+        from pyresample.geometry import AreaDefinition
```
Comment (djhoese) [inline pycoast/cw_base.py:1014]: Pyresample is not technically a dependency of pycoast. How do other methods handle this type of logic? I know the lon/lat -> x/y coordinates could be done with only Proj, but not to x/y grid indexes.

I'm ok with pyresample being depended on here, but would prefer there by a try/except ImportError and a message about points requiring pyresample.

File: pycoast/tests/test_pycoast.py [hunk 14]
```diff
@@ -677,31 +782,59 @@ def test_add_shapefile_shapes_agg(self):
...
-    @unittest.skip("All kwargs are not supported, so can't create equal results")
+#    @unittest.skip("All kwargs are not supported, so can't create equal results")
     def test_config_file_coasts_and_grid(self):
         from pycoast import ContourWriterAGG
         from pyresample.geometry import AreaDefinition
         overlay_config = os.path.join(os.path.dirname(__file__),
                                       "coasts_and_grid_agg.ini")
         grid_img = Image.open(os.path.join(os.path.dirname(__file__),
-                                           'grid_nh_agg.png'))
+                                           'grid_nh_cfg_agg.png'))
```
Comment (djhoese) [inline pycoast/tests/test_pycoast.py:792]: Does this mean this test never actually worked? Why doesn't this test need to be skipped now?

File: pycoast/cw_base.py (hunk 13 above)
Comment (yufeizhu600) [reply pycoast/cw_base.py:1014]: The reason I add the import and the isinstance() check is because other functions, except add_cities(), support two types of area_def, like

> area_def : (proj_str, area_extent) or AreaDefinition

However, add_cities() and add_points() only support area_def parameter as AreaDefinition object. In add_cities() there is a try/except when use get_xy_from_lonlat(), but I think "Explicit is better than implicit", so I add the type check at the beggining of the function, which needs a explicit import of pyresample. I will add a try/except to that.

File: pycoast/tests/test_pycoast.py (hunk 14 above)
Comment (yufeizhu600) [reply pycoast/tests/test_pycoast.py:792]: There was a bug in the .ini file for that unit test, which cause an exception without any image been generated. I think that is why it has had been skipped. I fixed the .ini file and the test can generate image from config file but I need to generate a new source image to compare with.

File: docs/source/points.rst (hunk 4 above)
Comment (yufeizhu600) [reply docs/source/points.rst:20]: Thanks for spell checking. It is done.

File: docs/source/points.rst (hunk 5 above)
Comment (yufeizhu600) [reply docs/source/points.rst:7]: Done.

File: docs/source/points.rst (hunk 6 above)
Comment (yufeizhu600) [reply docs/source/points.rst:5]: Done.

File: docs/source/points.rst (hunk 5 above)
Comment (yufeizhu600) [reply docs/source/points.rst:7]: done.

File: docs/source/points.rst (hunk 7 above)
Comment (yufeizhu600) [reply docs/source/points.rst:37]: Done.

File: docs/source/points.rst (hunk 8 above)
Comment (yufeizhu600) [reply docs/source/points.rst:38]: Done

File: docs/source/points.rst (hunk 9 above)
Comment (yufeizhu600) [reply docs/source/points.rst:47]: Done.

File: docs/source/points.rst (hunk 10 above)
Comment (yufeizhu600) [reply docs/source/points.rst:48]: Done.

File: docs/source/points.rst (hunk 11 above)
Comment (yufeizhu600) [reply docs/source/points.rst:64]: Yes, you are right, it was left over. My intention was pointing users to the docstrings as the reference of the parameters after showing the examples.

**Comment by yufeizhu600** [comment]
I have made the changes to the name of the textbox parameters. Regarding the images, I have only added new images for my new tests and the one was skipped, I cannot make it work with the image in the test suite, I have to add a new one, hope it is OK with you.  
Updates has been pushed, please check it out. Thanks.

**Comment by mraspaud** [comment]
Thanks for the addition! I think the next step is to factorize `add_cities` and `add_points` as one is a special case of the other, right ? But that will be another PR, I'm merging this as it is now.
